### ISSUE TITLE
fatal error caused by shell type

### ISSUE THREAD

**Issue by wanopt** [ansible/ansible#13459]
on mac 10.9.5, ansible versions > 1.9.4 has an error caused by default shell on remote freebsd 10.2 guest. 

default ssh to freebsd seems to use csh and thus give the error as below. 1.9.4 works fine since it does not generate the echo portion like

```
echo "$( echo $HOME/.ansible/tmp/ansible-tmp-1449517181.33-230267840450118 )" ). 

<127.0.0.1> SSH: EXEC ssh -C -vvv -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o Port=2222 -o 'IdentityFile="/Users/t/.vagrant.d/insecure_private_key"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=vagrant -o ConnectTimeout=30 -o ControlPath=/Users/t/.ansible/cp/ansible-ssh-%h-%p-%r -tt 127.0.0.1 ( umask 22 && mkdir -p "$( echo $HOME/.ansible/tmp/ansible-tmp-1449517181.33-230267840450118 )" && echo "$( echo $HOME/.ansible/tmp/ansible-tmp-1449517181.33-230267840450118 )" )

vagrant@server:~ % umask 22 && mkdir -p "$( echo $HOME/.ansible/tmp/ansible-tmp-1449517181.33-230267840450118
Illegal variable name.
```

configured ssh shell on vagrant and inside ansible privisioner through ansible_shell_type but still got the error.

**Comment by bcoca** [comment]
Can You Help Us Out?
====================

Thanks for filing a ticket! I am the friendly GitHub Ansibot.

It looks like you might not have filled out the issue description based on our standard issue template.  You might not have known about that, and that's ok too, we'll tell you how to do it.

We have a standard template because Ansible is a really busy project and it helps to have some standard information in each ticket, and GitHub doesn't yet provide a standard facility to do this like some other bug trackers.  We hope you understand as this is really valuable to us!.

Solving this is simple: please copy the contents of this [template](https://raw.githubusercontent.com/ansible/ansible/devel/ISSUE_TEMPLATE.md) and **paste it into the description** of your ticket.  That's it!

If You Had A Question To Ask Instead
====================================

If you happened to have a "how do I do this in Ansible" type of question, that's probably more of a user-list question than a bug report, and you should probably ask this question on the project mailing list instead.

However, if you think you have a bug, the report is the way to go!  We definitely want all the bugs filed :)  Just trying to help!

About Priority Tags
===================

Since you're here, we'll also share some useful information at this time.

In general tickets will be assigned a priority between P1 (highest) and P5, and then worked in priority order. We may also have some follow up questions along the way, so keeping up with follow up comments via GitHub notifications is a good idea.

Due to large interest in Ansible, humans may not comment on your ticket immediately.

Mailing Lists
=============

If you have concerns or questions, you're welcome to stop by the ansible-project or ansible-development mailing lists, as appropriate.  Here are the links:

  * https://groups.google.com/forum/#!forum/ansible-project - for discussion of bugs and how-to type questions
  * https://groups.google.com/forum/#!forum/ansible-devel - for discussion on how to implement a code change, or feature brainstorming among developers

Thanks again for the interest in Ansible!

**Comment by wanopt** [comment]
A temporary workaround for vagrant with freebsd 10.2 guest to work with ansible 2.0.0: manually changing the default shell of freebsd from csh to sh through vagrant builtin privision. 

config.vm.define :server do |server|
....
    server.vm.provision "shell",
      run: "always",
      inline: "chsh -s /bin/sh vagrant"
....
end

Something has changed at "mkdtemp" method for ansible versions > 1.9.4. The echo part echo "$( echo $HOME/.ansible/tmp/ansible-tmp-1449517181.33-230267840450118 )" ) from this method seems to have caused the problem for csh. Have not got time to dig into it though.

**Comment by bcoca** [comment]
have you set ansible_shell_type=csh for that host?

**Comment by wanopt** [comment]
No, I did not set ansible_shell_type=csh. I found that the worse part is that even when I make it NOT to be csh (e.g., sh), it seemed to have no effect. That made me to resort to vagrant provision instead.

**Comment by bcoca** [comment]
if your target machine is using a csh type shell you need to set that var, otherwise ansible will use the default (sh) which will produce the error you show above.

if you don't want to set the var then changing the shell to sh is the other way to make it work.

Closing this as it seems to just be a misunderstanding on the shell settings.

**Comment by wanopt** [comment]
Does version 1.9.4 work with csh out of box?

**Comment by wanopt** [comment]
In my test, 1.9.4 works with the target machine and 2.0 does not. I see some changes in "mkdtemp" method for ansible versions > 1.9.4 causes the problem.

**Comment by bcoca** [comment]
tested latest stable-2.0 and it works, there were versions of 2.0 in which this was broken, can you use latest?

**Comment by wanopt** [comment]
Looks like the method below (from version > 1.9.4) caused the problem. It issued a command like `umask 22 && mkdir -p "$( echo $HOME/.ansible/tmp/ansible-tmp-1449517181.33-230267840450118` through ssh. The command gives the error `Illegal variable name.` if being run under csh, same with the error reported from ansible. I checked 1.9.4 and the generated command from `mkdtemp`(1.9.4) is different and seems to able to succeed on csh. The method I referred to is in lib/ansible/plugins/shell/sh.py, which has been changed from 1.9.4 to 2.0.0. You may have the reason to change it. I just noticed that this change make it fail on csh.

```
    def mkdtemp(self, basefile=None, system=False, mode=None):
        if not basefile:
            basefile = 'ansible-tmp-%s-%s' % (time.time(), random.randint(0, 2**48))
        basetmp = self.join_path(C.DEFAULT_REMOTE_TMP, basefile)
        if system and (basetmp.startswith('$HOME') or basetmp.startswith('~/')):
            basetmp = self.join_path('/tmp', basefile)
        cmd = 'mkdir -p %s echo %s %s' % (self._SHELL_SUB_LEFT, basetmp, self._SHELL_SUB_RIGHT)
        cmd += ' %s echo %s echo %s %s' % (self._SHELL_AND, self._SHELL_SUB_LEFT, basetmp, self._SHELL_SUB_RIGHT)

        # change the umask in a subshell to achieve the desired mode
        # also for directories created with `mkdir -p`
        if mode:
            tmp_umask = 0o777 & ~mode
            cmd = '%s umask %o %s %s %s' % (self._SHELL_GROUP_LEFT, tmp_umask, self._SHELL_AND, cmd, self._SHELL_GROUP_RIGHT)

        return cmd

```

**Comment by bcoca** [comment]
pushed fix, this now works in stable-2.0 and devel also

**Comment by wanopt** [comment]
Thanks!
